### PR TITLE
[JENKINS-46175] Make workspace and launcher optional for CoreStep and CoreWrapperStep.

### PR DESCRIPTION
**Pull request by Zastai**
This also includes the same `PwdStep` spotbugs fix as https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/116 does.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -64,7 +64,8 @@
     <properties>
         <revision>2.21</revision>
         <changelist>-SNAPSHOT</changelist>
-        <jenkins.version>2.164.3</jenkins.version>
+        <jenkins.version>2.241-rc30039.aafbea2368b4</jenkins.version>
+ <!-- https://github.com/jenkinsci/jenkins/pull/4784 -->
```
Comment (jglick) [inline pom.xml:68]:
```suggestion
        <jenkins.version>2.241-rc30039.aafbea2368b4</jenkins.version> <!-- https://github.com/jenkinsci/jenkins/pull/4784 -->
```

File: src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java [hunk 2]
```diff
@@ -92,9 +93,26 @@ public SimpleBuildWrapper getDelegate() {
...
-            delegate.setUp(c, run, getContext().get(FilePath.class), getContext().get(Launcher.class), getContext().get(TaskListener.class), getContext().get(EnvVars.class));
+            // Get the context objects that may be required
+            final StepContext context = this.getContext();
+            final Run<?,?> run = context.get(Run.class);
+            final FilePath workspace = context.get(FilePath.class);
+            final Launcher launcher = context.get(Launcher.class);
+            final TaskListener listener = context.get(TaskListener.class);
+            final EnvVars env = context.get(EnvVars.class);
+            // Ensure those that are required are provided
+            Objects.requireNonNull(run);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java:104]: Fine but unnecessary: `getRequiredContext` handles this.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java [hunk 3]
```diff
@@ -148,8 +166,23 @@ private void doStart() throws Exception {
...
+            final Launcher launcher = context.get(Launcher.class);
+            final TaskListener listener = context.get(TaskListener.class);
+            // Ensure those that are required are provided
+            Objects.requireNonNull(run);
+            if (this.disposer.requiresWorkspace() && workspace == null) {
+                throw new MissingContextVariableException(FilePath.class);
+            }
+            if (this.disposer.requiresLauncher() && launcher == null) {
+                throw new MissingContextVariableException(Launcher.class);
+            }
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java:182]: As noted in https://github.com/jenkinsci/jenkins/pull/4784#discussion_r445877019 this whole `finished` method could just be reverted to the original.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java [hunk 4]
```diff
@@ -95,7 +97,13 @@ public boolean isTmp() {
 
         @Override protected String run() throws Exception {
             FilePath cwd = getContext().get(FilePath.class);
-            return (tmp ? WorkspaceList.tempDir(cwd) : cwd).getRemote();
+            Objects.requireNonNull(cwd);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java:100]: redundant

File: src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java [hunk 5]
```diff
@@ -95,7 +97,13 @@ public boolean isTmp() {
...
+            if (tmp) {
+                cwd = WorkspaceList.tempDir(cwd);
+                if (cwd == null)
+                    throw new IOException("Failed to set up a temporary directory.");
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java:104]: (also in #116)

Comment (Zastai) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java:104]: I know - but until that gets merged (and I merge in master here), not applying the same fix here caused spotbugs failures.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java (hunk 4 above)
Comment (Zastai) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java:100]: Yes, but non in a way that is obvious from static analysis.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java (hunk 2 above)
Comment (Zastai) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java:104]: Yes, but non in a way that is obvious from static analysis.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java (hunk 4 above)
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java:100]: If SpotBugs is complaining, you can always use `@SuppressFBWarnings` or simply

```java
assert cwd != null;
```

File: src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java (hunk 5 above)
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/PwdStep.java:104]: Right, just noting that this is an accommodation to newer `jenkins.version` generally.

File: pom.xml (hunk 1 above)
Comment (Zastai) [reply pom.xml:68]: Was intended to be on the same line. Will adjust this when I update this PR to use a newer incremental.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java (hunk 2 above)
Comment (Zastai) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java]: Adjusted to use asserts instead.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java (hunk 3 above)
Comment (Zastai) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java]: Updated to a recent incremental, so these are gone

File: pom.xml (hunk 1 above)
Comment (Zastai) [reply pom.xml]: Now done

**Comment by jglick** [comment]
BTW force pushes prevent reviewers from using the GH feature of reviewing only new changes to a PR—you have to start from scratch.

**Comment by Zastai** [comment]
Yes, that was my bad. I had squashed a commit, but had previously pushed a part of that, so needed to force-push.

**Comment by jglick** [comment]
Not a problem, just FYI.

**Comment by jglick** [comment]
Is there an alternate implementation based on https://github.com/jenkinsci/jenkins/pull/4829 for comparison?

**Comment by Zastai** [comment]
Not yet. I started a branch but will need to update it to the current state of that PR. And then wait for an incremental to make the build work.

**Comment by Zastai** [comment]
Converted to draft, because there are two alternative core PRs (and soon 2 PRs here too).

**Comment by Zastai** [comment]
Closing - https://github.com/jenkinsci/jenkins/pull/4784 has been closed, and https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/123 replaces this PR.

**Pull request closed without merging**